### PR TITLE
Enforce maximum number of ads displayed across all user profiles

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/9805

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/handbook/blob/master/development/security.md#when-is-a-security-review-needed), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

Confirm ad notifications, promoted content ads and new tab page ads do not exceed the per hour and per day caps across multiple user profiles

### CODE REVIEW

File: vendor/brave-ios/Ads/BATBraveAds.mm [hunk 1]
```diff
@@ -1016,6 +1026,21 @@ - (void)closeNotification:(const std::string &)id
...
+  if (adEventHistory) {
+    adEventHistory->Record(ad_type, confirmation_type, timestamp);
+  }
+}
+
+- (std::vector<uint64_t>)getAdEvents:(const std::string&)ad_type
+                    confirmationType:(const std::string&)confirmation_type {
+  if (adEventHistory) {
+    return adEventHistory->Get(ad_type, confirmation_type);
+  }
```
Comment (kylehickinson) [inline vendor/brave-ios/Ads/BATBraveAds.mm:1041]: Won't compile, missing return statement for nil `adEventHistory`:

```suggestion
  if (adEventHistory) {
    return adEventHistory->Get(ad_type, confirmation_type);
  }
  return {};
```

**Review by kylehickinson — CHANGES_REQUESTED** [review]
If the only use of `AdEventHistory` is to handle `AdsClient` methods, why is it up to the client to handle?

Is it because of desktop/Android needing different event history for different profiles? Wouldn't they have entirely different `Ads` setups for each profile?

File: vendor/brave-ios/Ads/BATBraveAds.mm (hunk 1 above)
Comment (tmancey) [reply vendor/brave-ios/Ads/BATBraveAds.mm]: Fixed

**Comment by tmancey** [comment]
I will DM you to discuss

**Review by kylehickinson — APPROVED** [review]

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/frequency_capping.cc [hunk 2]
```diff
@@ -0,0 +1,34 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "bat/ads/internal/frequency_capping/frequency_capping.h"
```
Comment (moritzhaller) [inline vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/frequency_capping.cc:6]: @tmancey delete test file

**Review by moritzhaller — APPROVED** [review]
LGTM++

File: vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/frequency_capping.cc (hunk 2 above)
Comment (tmancey) [reply vendor/bat-native-ads/src/bat/ads/internal/frequency_capping/frequency_capping.cc]: Fixed :-)

**Pull request merged**